### PR TITLE
[New sample] Open Chrome API reference page (omnibox, alarms, messaging sample)

### PR DESCRIPTION
**Pull request by AmySteam**
## User Story
As a web developer, I want to open extension API reference pages on Chrome, so that I can quickly learn how to use a specific extension API.

**Bonus**: A daily extension tip will be displayed as a popover after clicking the "tip" navigation list item.
(Currently, it requires Canary build and the "Experimental Web Platform features" flag enabled. The release date was moved to M114, so this approach may need to be abandoned if we want to launch before then.)

### How to use it?
1. Enter the keyword "api" in the browser address bar.
2. Then, press "tab" or "space".
3. Start typing they API keyword.
4. A list the most common APIs will show or a list of past searches
5. Select the API desired, and a new page will open to the Chrome API reference page.

### TODO:
- [ ]  The [Popover API](https://chromestatus.com/feature/5463833265045504) is currently being developed. If it's not ready by the launch, then we will make adjustments. The popover was chosen to simplify the example.
- [ ] The ReadME.md will have a link to the tutorial

### CODE REVIEW

File: functional-samples/tutorial.open-api-reference/content.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Popover API https://chromestatus.com/feature/5463833265045504
+
+const nav = document.querySelector('.navigation-rail__links');
+
+chrome.runtime.sendMessage({ greeting: 'tip' }).then(({ tip }) => {
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/content.js:5]: Would be great to consistently use the more modern async/await

File: functional-samples/tutorial.open-api-reference/manifest.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  "manifest_version": 3,
+  "name": "Open extension API docs",
+  "version": "1.0.0",
+  "icons": {
+    "16": "icon-16.png",
+    "128": "icon-128.png"
+  },
+  "background": {
+    "service_worker": "sw.js",
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/manifest.json:10]: suggestion: rename to service-worker.js for folks who're not aware of the sw abbreviation

File: functional-samples/tutorial.open-api-reference/sw-api-list.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
+const apiList = [
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-api-list.js:1]: suggestion: move to a separate JSON file

File: functional-samples/tutorial.open-api-reference/sw-api-list.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      'Use the <match>TopSites API</match> to access the most visited sites that are displayed on the new tab page.'
+  },
+  {
+    content: 'webNavigation',
+    description:
+      'Use the <match>WebNavigation API</match> to receive notifications about the status of navigation requests in-flight.'
+  }
+];
+
+export async function getAPIsuggestions(input) {
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-api-list.js:83]: nit: rename to `getApiSuggestions`

File: functional-samples/tutorial.open-api-reference/sw-api-list.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  const result = [];
+  for (const api of apiList) {
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-api-list.js:85]: tip: a quicker way to write this is:

result = apiList.filter(api => api.content.startsWith(input))

File: functional-samples/tutorial.open-api-reference/sw-api-list.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      description: 'Matching Chrome APIs'
+    });
+    console.log('result', result);
+    return result;
+  } else {
+    await chrome.omnibox.setDefaultSuggestion({
+      description: 'No matches found. Choose from past searches'
+    });
+    // Retrieve defaults if keyword has no matches
+    const { apiSugg } = await chrome.storage.local.get('apiSugg');
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-api-list.js:103]: nit: s/apiSugg/apiSuggestions   - I'd avoid abbrevations as it makes it harder to comprehend the code

File: functional-samples/tutorial.open-api-reference/sw-api-list.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  const result = [];
+  for (const api of apiList) {
+    if (api.content.startsWith(input)) {
+      result.push(api);
+      console.log('pushing api', api);
+    }
+  }
+
+  if (result.length) {
+    await chrome.omnibox.setDefaultSuggestion({
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-api-list.js:93]: I find this a bit unexpected for a method called getApiSuggestions to also update the omnibox. For better separation of concerns I'd split these up into two steps

File: functional-samples/tutorial.open-api-reference/sw-omnibox.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+// Displays the suggestions after user starts typing
+chrome.omnibox.onInputChanged.addListener(async (input, suggest) => {
+  const suggestions = await getAPIsuggestions(input);
+  suggest(suggestions);
+});
+
+// Opens the reference page of the chosen API
+chrome.omnibox.onInputEntered.addListener(async (input) => {
+  await chrome.tabs.create({ url: chromeURL + input });
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-omnibox.js:22]: Do we need to await here?

File: functional-samples/tutorial.open-api-reference/sw-omnibox.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  // Saves the latest keyword
+  updateHistory(input);
+});
+
+async function updateHistory(input) {
+  try {
+    const { apiSugg } = await chrome.storage.local.get('apiSugg');
+    apiSugg.unshift(input);
+    apiSugg.splice(4);
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-omnibox.js:31]: nit: magic number - it'd be good to encode this in a constant named e.g. NUMBER_OF_PREVIOUS_SEARCHES

File: functional-samples/tutorial.open-api-reference/sw-omnibox.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+// Opens the reference page of the chosen API
+chrome.omnibox.onInputEntered.addListener(async (input) => {
+  await chrome.tabs.create({ url: chromeURL + input });
+  // Saves the latest keyword
+  updateHistory(input);
+});
+
+async function updateHistory(input) {
+  try {
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-omnibox.js:28]: Is the try catch handler needed here?

File: functional-samples/tutorial.open-api-reference/sw-tips.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
+console.log('sw-tips.js');
+
+// Fetch tip & save in storage
+const updateTip = async () => {
+  const response = await fetch('https://extension-tips.glitch.me/tips.json');
+  const tips = await response.json();
+  const index = Math.floor(Math.random() * tips.length);
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-tips.js:7]: suggestion: rename to `randomIndex`

File: functional-samples/tutorial.open-api-reference/sw-tips.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  }
+});
+
+// Retrieve tip of the day
+chrome.alarms.onAlarm.addListener(updateTip);
+
+// Send tip to content script via messaging
+chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
+  if (message.greeting === 'tip') {
+    chrome.storage.local.get('tip').then(sendResponse);
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-tips.js:25]: Please use async await

**Review by sebastianbenz — COMMENTED** [review]
This looks great - just a super quick first pass with a few nits.

File: functional-samples/tutorial.open-api-reference/sw-api-list.js (hunk 3 above)
Comment (AmySteam) [reply functional-samples/tutorial.open-api-reference/sw-api-list.js:1]: I think we are already giving an example of how to fetch with the extension tips feature. I think, if SW supported [Import assertions](https://chromestatus.com/feature/5765269513306112), I would be more open to moving this into a separate JSON file.

How about I move it into a separate JS file? 🧐😁

File: functional-samples/tutorial.open-api-reference/sw-tips.js (hunk 12 above)
Comment (AmySteam) [reply functional-samples/tutorial.open-api-reference/sw-tips.js:25]: I used `.then()` here b/c the return value of the `onMessage` listener determines if Chrome expects `sendResponse` to be called.

If we use async/await, this function will return a Promise, and Chrome will always expect `sendResponse` to be called.

We could wrap the call to `.get()` in an async IIFE, but it seems like overkill for this tutorial😊

**Comment by AmySteam** [comment]
@sebastianbenz Thanks for the review. I applied most of your suggestions. I wrote comments on the suggestions I had a few thoughts on.

File: functional-samples/tutorial.open-api-reference/sw-suggestions.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import apiList from './api-list.js';
+
+/**
+ * Returns a list of suggestions and a description for the default suggestion
+ */
+export async function getApiSuggestions(input) {
+  const filtered = apiList.filter((api) => api.content.startsWith(input));
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-suggestions.js:7]: suggestion: rename to better explain what this variable includes
```suggestion
  const suggestions = apiList.filter((api) => api.content.startsWith(input));
```

File: functional-samples/tutorial.open-api-reference/sw-tips.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
+console.log('sw-tips.js');
```
Comment (sebastianbenz) [inline functional-samples/tutorial.open-api-reference/sw-tips.js:1]: remove?

File: functional-samples/tutorial.open-api-reference/sw-tips.js (hunk 12 above)
Comment (sebastianbenz) [reply functional-samples/tutorial.open-api-reference/sw-tips.js:25]: I didn't know that, have you got a link to where this is explained. Curious to learn how this works.

**Review by sebastianbenz — COMMENTED** [review]
Thanks! Left a few more comments, only nits.

File: functional-samples/tutorial.open-api-reference/sw-suggestions.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+  console.log('filtered', filtered);
```
[the remark this answers is not in the document]
Comment (AmySteam) [reply functional-samples/tutorial.open-api-reference/sw-suggestions.js:8]: removed

File: functional-samples/tutorial.open-api-reference/sw-omnibox.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
+import { getApiSuggestions } from './sw-suggestions.js';
+
+console.log('sw-omnibox.js');
```
[the remark this answers is not in the document]
Comment (AmySteam) [reply functional-samples/tutorial.open-api-reference/sw-omnibox.js:3]: I was using these to show where to find the logs for the service worker early on in the tutorial.

File: functional-samples/tutorial.open-api-reference/sw-tips.js (hunk 14 above)
Comment (AmySteam) [reply functional-samples/tutorial.open-api-reference/sw-tips.js:1]: I was using it to show where to find the logs for the service worker in the initial steps of the tutorial.

File: functional-samples/tutorial.open-api-reference/sw-tips.js (hunk 12 above)
Comment (sebastianbenz) [reply functional-samples/tutorial.open-api-reference/sw-tips.js:25]: Thanks ! That's not intuitive. 

One thing: maybe rename the message field from `greeting` to `type` as it better expresses what this field is used for.

**Review by sebastianbenz — APPROVED** [review]
This looks good, once https://github.com/GoogleChrome/chrome-extensions-samples/pull/848/files#r1144016971 is fixed.

**Pull request merged**